src/move_basic.cpp

```diff
@@ -637,151 +600,85 @@ bool MoveBasic::rotate(double yaw, const std::string& drivingFrame)
 
         if (actionServer->isPreemptRequested()) {
             ROS_INFO("MoveBasic: Stopping rotation due to preempt");
-            done = true;
-            velocity = 0;
+            sendCmd(0, 0);
+            actionServer->setPreempted();
+            return false;
         }
 
-        //ROS_INFO("%f %f %f", rad2deg(angleRemaining), angleRemaining, velocity);
-
         if (std::abs(angleRemaining) < angularTolerance || oscillations > 2) {
+            ROS_INFO("MoveBasic: Done rotation, error %f degrees", rad2deg(angleRemaining));
             velocity = 0;
             done = true;
-            ROS_INFO("MoveBasic: Done rotation, error %f degrees", rad2deg(angleRemaining));
         }
+
+        bool counterwise = (angleRemaining < 0.0);
+        if (counterwise) {
+            velocity = -velocity;
+        }
+
         sendCmd(velocity, 0);
+        ROS_DEBUG("Angle remaining: %f, Angular velocity: %f", rad2deg(angleRemaining), velocity);
     }
     return done;
 }
 
 // Move forward specified distance
 
 bool MoveBasic::moveLinear(tf2::Transform& goalInDriving,
-                           const std::string& planningFrame,
                            const std::string& drivingFrame)
 {
-    bool done = false;
-    ros::Rate r(50);
-
-    bool waitingForObstacle = false;
-    int  waitingLoops = 0;
-    double forwardObstacleThreshold = 1.5;  // if distance <  velocity times this we stop
-    ros::Time obstacleTime;
-
-    tf2::Transform poseDrivingInitial;
-    if (!getTransform(baseFrame, drivingFrame, poseDrivingInitial)) {
-         abortGoal("MoveBasic: Cannot determine robot pose for linear");
-         return false;
-    }
-
-    tf2::Vector3 linear = (poseDrivingInitial.getOrigin() -
-                           goalInDriving.getOrigin());
-    linear.setZ(0);
-    double requestedDistance = linear.length();
-
     tf2::Transform poseDriving;
     if (!getTransform(drivingFrame, baseFrame, poseDriving)) {
          abortGoal("MoveBasic: Cannot determine robot pose for linear");
          return false;
     }
-
     tf2::Transform goalInBase = poseDriving * goalInDriving;
     tf2::Vector3 remaining = goalInBase.getOrigin();
-    bool forward = (remaining.x() > 0);
+
+    remaining.setZ(0);
+    double requestedDistance = remaining.length();
+
+    bool pausingForObstacle = false;
+    ros::Time obstacleTime;
+    ros::Duration runawayTimeout(runawayTimeoutSecs);
```

Comment:
It's interesting this was done in terms of time.  Does it work well?

Reply:
I added a timeout, since localization with LiDAR is jumping +-0.02 and when we get really close to goal, the robot might thought it is over the goal(cos(angleRemaining) becomes negative since the goal is behind - at least what the localization think it is)). If there where no timeout, goal would be immediately aborted as soon as the localization would jump "over the goal". This happened often when testing(of course when we where trying to reach the goal better than the localization accuracy).

Reply:
I think that's an argument towards retrying a goal instead of just aborting it, so that the goal is reached in any case and merely triggers a recalculation when the angle doesn't seem to be sane.